libcudacxx/include/cuda/std/__atomic/types/base.h

```diff
@@ -0,0 +1,232 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___ATOMIC_TYPES_BASE_H
+#define _LIBCUDACXX___ATOMIC_TYPES_BASE_H
+
+#include <cuda/std/detail/__config>
+
+#include <cuda/std/__atomic/functions.h>
+#include <cuda/std/__atomic/types/common.h>
+#include <cuda/std/__type_traits/is_trivially_copyable.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <typename _Tp>
+struct __atomic_storage
+{
+  using __underlying_t                = _Tp;
+  static constexpr __atomic_tag __tag = __atomic_tag::__atomic_base_tag;
+
+#if !defined(_CCCL_COMPILER_GCC) || (__GNUC__ >= 5)
+  static_assert(is_trivially_copyable<_Tp>::value, "std::atomic<Tp> requires that 'Tp' be a trivially copyable type");
```

Comment:
We want to move towards inline variables when possible:
```suggestion
  static_assert(_LIBCUDACXX_TRAIT(is_trivially_copyable, _Tp), "std::atomic<Tp> requires that 'Tp' be a trivially copyable type");
```